src/udev/udev-netlink.c

```diff
@@ -85,3 +87,112 @@ int link_info_get(sd_netlink **rtnl, int ifindex, LinkInfo *ret) {
+
+        if (device_get_cached_sysattr_value(device, "type", NULL) == -ENODATA) {
+                _cleanup_free_ char *str = NULL;
+
+                if (asprintf(&str, "%"PRIu16, info->iftype) < 0)
+                        return -ENOMEM;
+
+                r = device_cache_sysattr_value(device, "type", str);
+                if (r < 0)
+                        return r;
+
+                TAKE_PTR(str);
+        }
+
+        if (device_get_cached_sysattr_value(device, "address", NULL) == -ENODATA) {
+                _cleanup_free_ char *str = NULL;
+
+                str = new(char, HW_ADDR_TO_STRING_MAX);
+                if (!str)
+                        return -ENOMEM;
+
+                r = device_cache_sysattr_value(device, "address", hw_addr_to_string(&info->hw_addr, str));
+                if (r < 0)
+                        return r;
+
+                TAKE_PTR(str);
+        }
+
+        if (device_get_cached_sysattr_value(device, "iflink", NULL) == -ENODATA) {
+                _cleanup_free_ char *str = NULL;
+
+                if (asprintf(&str, "%"PRIu32, info->iflink) < 0)
+                        return -ENOMEM;
+
+                r = device_cache_sysattr_value(device, "iflink", str);
+                if (r < 0)
+                        return r;
+
+                TAKE_PTR(str);
+        }
+
+        if (info->support_phys_port_name &&
+            device_get_cached_sysattr_value(device, "phys_port_name", NULL) == -ENODATA) {
+                _cleanup_free_ char *str = NULL;
+
+                if (info->phys_port_name) {
+                        str = strdup(info->phys_port_name);
+                        if (!str)
+                                return -ENOMEM;
+                }
+
+                r = device_cache_sysattr_value(device, "phys_port_name", str);
+                if (r < 0)
+                        return r;
+
+                TAKE_PTR(str);
+        }
+
+        return 0;
+}
+
+int device_get_sysattr_value_maybe_from_netlink(
+                sd_device *device,
+                sd_netlink **rtnl,
+                const char *sysattr,
+                const char **ret_value) {
+
+        _cleanup_(link_info_clear) LinkInfo info = LINK_INFO_NULL;
+        int ifindex, r;
+
+        assert(device);
+        assert(rtnl);
+        assert(sysattr);
+
+        if (sd_device_get_ifindex(device, &ifindex) < 0)
+                return sd_device_get_sysattr_value(device, sysattr, ret_value);
+
+        if (!STR_IN_SET(sysattr, "type", "address", "iflink", "phys_port_name"))
+                return sd_device_get_sysattr_value(device, sysattr, ret_value);
+
+        r = device_get_cached_sysattr_value(device, sysattr, ret_value);
+        if (r != -ENODATA)
+                return r;
+
+        r = link_info_get(rtnl, ifindex, &info);
+        if (r < 0)
+                return r;
+
+        r = device_cache_sysattr_from_link_info(device, &info);
+        if (r < 0)
+                return r;
```

Comment:
wouldn't it be better to reorder:

```c

r = device_get_cached_sysattr_value(device, sysattr, ret_value);
if (r != -ENODATA)
        return r;
if (!STR_IN_SET(sysattr, "type", "address", "iflink", "phys_port_name"))
        return r;
 if (sd_device_get_ifindex(device, &ifindex) < 0)
        return r;

r = link_info_get(rtnl, ifindex, &info);
if (r < 0)
        return r;
…
```

Reply:
I'd like to make this function mostly a light wrapper of sd_device_get_sysattr_value() for non network interface devices.
I guess `sd_device_get_ifindex()` is the fastest among the above three.
Also, `device_get_cached_sysattr_value()` will be called in `sd_device_get_sysattr_value()`, so to avoid it called twice, it should be called at last.

So, I'd like to keep the current order.